src/Microsoft.Diagnostics.Monitoring.Egress.AzureStorage/AzureBlobEgressEndpoint.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress.AzureStorage
+{
+    internal class AzureBlobEgressEndpoint : IEgressEndpoint<AzureBlobEgressStreamOptions>
+    {
+        private readonly AzureBlobEgressEndpointOptions _settings;
+
+        public AzureBlobEgressEndpoint(AzureBlobEgressEndpointOptions settings)
+        {
+            _settings = settings;
+        }
+
+        public async Task<EgressResult> EgressAsync(
+            Func<CancellationToken, Task<Stream>> action,
+            string name,
+            AzureBlobEgressStreamOptions options,
+            CancellationToken token)
+        {
+            using var stream = await action(token);
+
+            return await Upload(stream, name, options, token);
+        }
+
+        public async Task<EgressResult> EgressAsync(
+            Func<Stream, CancellationToken, Task> action,
+            string name,
+            AzureBlobEgressStreamOptions options,
+            CancellationToken token)
+        {
+            using var stream = new MemoryStream();
+
+            await action(stream, token);
+
+            stream.Position = 0;
+
+            return await Upload(stream, name, options, token);
+        }
+
+        private async Task<EgressResult> Upload(
+            Stream stream,
+            string fileName,
+            AzureBlobEgressStreamOptions options,
+            CancellationToken token)
+        {
+            var serviceUriBuilder = new UriBuilder()
```

Comment:
Have you looked at BlobUriBuilder?

Reply:
That class looks like it is used for decomposing an existing URI, allows you to update components of it, and recompose the URI. It does not seem to be good for composing a URI from scratch.